### PR TITLE
Message interface declutter & mention / component abstraction

### PR DESCRIPTION
**Pull request by arynxd**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This PR implements the new message mentions / components abstractions that are planned for v5. It will be draft until the checklist is complete. This is open for contribution.

- [x] Decide on nullability for getContentX
- [x] Decide on behavior on empty components / content 
- [ ] Decide on how this should impact internal Message abstractions (AbstractMessage / ReceivedMessage)
- [ ] Decide on the abstraction layout of Message(Mentions/Components), potential interface?

# TODO

- [ ] Make getContentX nonnull
- [ ] Throw on empty content access
- [ ] Expose application flags through JDA

### CODE REVIEW

**Comment by arynxd** [comment]
@DV8FromTheWorld would be good if you could comment on the task items.

File: src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            Guild guild, JDAImpl jda
+    ) {
+        this.content = content;
+        this.mentionsEveryone = mentionsEveryone;
+        this.userIds = userIds;
+        this.roleIds = roleIds;
+        this.jda = jda;
+        this.guild = guild;
+
+        this.users = Collections.unmodifiableList(processMentions(Message.MentionType.USER, new ArrayList<>(), true, this::matchUser));
```
Comment (Sanduhr32) [inline src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java:50]: You wrap a lot of collections returned by `processMentions()` with a `Collections.unmodifiableList()` call.
Maybe change the methods returned collection.
```suggestion
        this.users = processMentions(Message.MentionType.USER, new ArrayList<>(), true, this::matchUser);
```

File: src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            try
+            {
+                T elem = map.apply(matcher);
+                if (elem == null || (distinct && collection.contains(elem)))
+                    continue;
+                collection.add(elem);
+            }
+            catch (NumberFormatException ignored) {}
+        }
+        return collection;
```
Comment (Sanduhr32) [inline src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java:159]:
```suggestion
        return Collections.unmodifiableList(collection);
```

File: src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                        .map(guild::getMember)
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toList())
+            );
+        }
+
+        this.roles = Collections.unmodifiableList(processMentions(Message.MentionType.ROLE, new ArrayList<>(), true, this::matchRole));
+        this.textChannels = Collections.unmodifiableList(processMentions(Message.MentionType.CHANNEL, new ArrayList<>(), true, it -> matchChannel(it, ChannelType.TEXT, TextChannel.class)));
+        this.voiceChannels = Collections.unmodifiableList(processMentions(Message.MentionType.CHANNEL, new ArrayList<>(), true, it -> matchChannel(it, ChannelType.VOICE, VoiceChannel.class)));
+        this.emotes = Collections.unmodifiableList(processMentions(Message.MentionType.EMOTE, new ArrayList<>(), true, this::matchEmote));
```
Comment (Sanduhr32) [inline src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java:70]:
```suggestion
        this.roles = processMentions(Message.MentionType.ROLE, new ArrayList<>(), true, this::matchRole);
        this.textChannels = processMentions(Message.MentionType.CHANNEL, new ArrayList<>(), true, it -> matchChannel(it, ChannelType.TEXT, TextChannel.class));
        this.voiceChannels = processMentions(Message.MentionType.CHANNEL, new ArrayList<>(), true, it -> matchChannel(it, ChannelType.VOICE, VoiceChannel.class));
        this.emotes = processMentions(Message.MentionType.EMOTE, new ArrayList<>(), true, this::matchEmote);
```

File: src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        this.userIds = userIds;
+        this.roleIds = roleIds;
+        this.jda = jda;
+        this.guild = guild;
+
+        this.users = Collections.unmodifiableList(processMentions(Message.MentionType.USER, new ArrayList<>(), true, this::matchUser));
+
+        if (guild == null)
+        {
+            this.members = new ArrayList<>();
```
Comment (Sanduhr32) [inline src/main/java/net/dv8tion/jda/api/entities/MessageMentions.java:54]: Let's also wrap this.
```suggestion
            this.members = Collections.unmodifiableList(new ArrayList<>());
```

**Review by Sanduhr32 — COMMENTED** [review]
@arynxd you still working on this?

**Comment by Sanduhr32** [comment]
> This will also need to address mentions in slash command interactions. Options of type STRING can contain mentions, which are resolved by discord.

Should options of types user, channel, role & mentionable also be included in their respective lists?

**Comment by DV8FromTheWorld** [comment]
Superseded by #2015

**Pull request closed without merging**